keps/sig-apps/20181220-statefulset-volume-expansion.md

```diff
@@ -0,0 +1,103 @@
+
+It would be easier and expected to allow for changes to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+Relevant Issues:
+
+* https://github.com/kubernetes/kubernetes/issues/71477
+* https://github.com/kubernetes/kubernetes/issues/72198
+
+### Goals
+
+Allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet to propagate to all associated persistent volume claims.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+The apiserver will allow for increases to storage requests in the `volumeClaimTemplates` component of a StatefulSet. Additionally, for each volumeClaimTemplate being expanded, it will be necessary to check if its associated StorageClass has volume expansion enabled. This will be achieved by using an admission controller to check if the associated StorageClasses have `allowVolumeExpansion = true`.
+
+During the StatefulSet update process, the StatefulSet controller will detect an update to a `volumeClaimTemplate` by comparing the updated and current revision of the StatefulSet. This requires the `VolumeClaimTemplates` component of the StatefulSet to be recorded in the StatefulSet's `ControllerRevision` object. 
+
+While updating a pod, the StatefulSet controller will update a referenced persistent volume claim object if its storage request in the associated `volumeClaimTemplate` has been increased.
+
+Not all volumes support online control-plane expansion so this design aims to support both online and offline control-plane volume expansion as follows.
+
+Prerequisite: The StatefulSet controller must be able to determine if a volume supports online expansion.
+
+To accomplish this a new method `RequiresOfflineResize` will be added to the `ExpandableVolumePlugin` interface. Through this it will possible to determine which in-tree volume plugins support online expansion. 
+
+For volumes provisioned through the CSI, an optional boolean field `SupportsOnlineExpansion` will be added into the `CSIDriverSpec` field of the `CSIDriver` object. The `CSIDriver` object stores information about a CSI volume driver deployed on the cluster. If the `CSIDriver` object or the `SupportsOnlineExpansion` field is not registered for a particular volume provisioned through the CSI, the statefulset controller will assume the volume does not support online expansion. Adding this functionality mandates an API change for the mentioned objects.
+
+If a volume is deemed to support online expansion it will be expanded before the pod is terminated for the update. The StatefulSet controller will wait for the file system resize to complete on all such volumes before terminating the associated pod.
+
+Otherwise, the volume will be expanded offline after the pod is terminated and before it is recreated.
```

Comment:
Can you expand on this bit? Like - if volume supports only offline expansion then we will terminate the pod and expand the PVC and then recreate the pod? Do we also wait for file system expansion to finish after starting the pod?

For completeness sake can we document what happens when volume types are mixed?

Reply:
Sure I will expand on the flow and document what will occur when volume types are mixed with an example. Yes I was going to wait for file system expansion

Reply:
One thing that came up while discussing this KEP with @msau42 is - for offline only expansion if we are going to wait for PVC expansion to finish before starting the pods managed by statefulsets and lets say volume expansion failed then we may not be able to recover from it and we won't be able to start the pods.

Currently we do not have a "resizing failed" condition on PVC, so other than events and the fact that PV size is never updated, we may not get a clear indication that volume expansion has failed.  

This is a general issue with volume expansion though - https://github.com/kubernetes/kubernetes/issues/73036 which will be solved when resizing goes GA. But interaction with stateful set makes it kinda little worse because pods can't start at all, since stateful set controller will be waiting on resizing to finish.

Reply:
I would say that the update being stuck is fine vs completing the update with the volumes remaining unexpanded. Another thing is we can't let the controller roll back the pod to the previous revision since we don't allow for decreases in storage requests.

So even if the pod is brought back up it would technically be neither at the current or updated revision of the set. In this case it may be best to just leave it terminated and surface an indication of this to the statefulset.

Reply:
We still need to document how user can recover from this. We do not want an user with many TB of data to be not able to start his statefulset if resize wasn't working.  

For general resizing.. the documented workaround is:
1. Edit the PV to have "Retain" reclaim policy.
2. Delete the PVC.
3. Recreate the PVC which binds to PV that was before.
4. Restart the pod.

Will something like this work for statefulsets?

Reply:
yep 
1. Delete StatefulSet. PVC and PV preseved
2. _The same procedure above_
3. Recreate Statefulset with old spec